### PR TITLE
feat: [PIE-4294]: utility for expanding execution yaml for CI

### PR DESCRIPTION
**Pull request by sahilhindwani-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34823)

### CODE REVIEW

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 13 of 15 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34823)* status: 18 file(s) to review. 1 of 2 approvals obtained.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 6 of 15 files at r1, 1 of 5 files at r2, 13 of 13 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34823)* status: 13 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`320-ci-execution/src/main/java/io/harness/ci/integrationstage/InitializeStepGenerator.java` line 40 at r3](https://reviewable.io/reviews/harness/harness-core/34823#-N5mqodh1iAdjJZP5Jmx:-N5mqodh1iAdjJZP5Jmy:b-e7w91b) ([raw file](https://github.com/harness/harness-core/blob/0be695d4d233ad2a9fa34748de9337b933bd3a7f/320-ci-execution/src/main/java/io/harness/ci/integrationstage/InitializeStepGenerator.java#L40)):*
> ```Java
>     IntegrationStageConfig integrationStageConfig = IntegrationStageUtils.getIntegrationStageConfig(stageElementConfig);
> 
>     List<ExecutionWrapperConfig> openedExecutionElement = new ArrayList<>();
> ```

expanded?



_Code quote:_
```Java
opened
```
___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 34 at r3](https://reviewable.io/reviews/harness/harness-core/34823#-N5msKFM2cyHHTW7APuk:-N5msKFM2cyHHTW7APul:boaa02p) ([raw file](https://github.com/harness/harness-core/blob/0be695d4d233ad2a9fa34748de9337b933bd3a7f/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L34)):*
> ```Java
>       return Collections.singletonList(nodeWithStrategy);
>     }
>     StrategyConfig strategyConfig = JsonPipelineUtils.read(node.toString(), StrategyConfig.class);
> ```

Instead of reading complete object, maybe we can just check manually? the reason, lot of deserialisation, I agree that manual code would be there, but optimising code is also necessary for this, as its ci step, maybe you can create a util to check if jsonNode is matrix/for/parallel and that util can contain if StrategyConfig is passed



_Code quote:_
```Java
JsonPipelineUtils.read
```
___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 54 at r3](https://reviewable.io/reviews/harness/harness-core/34823#-N5mszo47HcJbzdzQphf:-N5mszo47HcJbzdzQphg:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/0be695d4d233ad2a9fa34748de9337b933bd3a7f/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L54)):*
> ```Java
>       if (executionWrapperConfig.getParallel() != null) {
>         ParallelStepElementConfig parallelStepElementConfig =
>             YamlUtils.read(executionWrapperConfig.getParallel().toString(), ParallelStepElementConfig.class);
> ```

same as above



_Code quote:_
```Java
YamlUtils.read
```
___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 68 at r3](https://reviewable.io/reviews/harness/harness-core/34823#-N5mt6RWDF7lm703Rjh5:-N5mt6RWDF7lm703Rjh6:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/0be695d4d233ad2a9fa34748de9337b933bd3a7f/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L68)):*
> ```Java
>       if (executionWrapperConfig.getStepGroup() != null) {
>         StepGroupElementConfig stepGroupElementConfig =
>             YamlUtils.read(executionWrapperConfig.getStepGroup().toString(), StepGroupElementConfig.class);
> ```

same as above



_Code quote:_
```Java
YamlUtils.read
```
___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 74 at r3](https://reviewable.io/reviews/harness/harness-core/34823#-N5mtDMXBK4hahPH5bEU:-N5mtDMXBK4hahPH5bEV:b-7mme34) ([raw file](https://github.com/harness/harness-core/blob/0be695d4d233ad2a9fa34748de9337b933bd3a7f/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L74)):*
> ```Java
>         }
>         stepGroupElementConfig.setSteps(executionWrapperConfigs);
>         JsonNode stepGroupNode = JsonPipelineUtils.asTree(stepGroupElementConfig);
> ```

can we check to see if without we can do?



_Code quote:_
```Java
sonPipelineUtils.asTree(
```
___
*[`878-ng-common-utilities/src/test/java/io/harness/steps/matrix/StrategyHelperTest.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/34823#-N5mtK2141ds1i-wnfPD:-N5mtK2141ds1i-wnfPE:b-8op02x) ([raw file](https://github.com/harness/harness-core/blob/0be695d4d233ad2a9fa34748de9337b933bd3a7f/878-ng-common-utilities/src/test/java/io/harness/steps/matrix/StrategyHelperTest.java#L29)):*
> ```Java
> 
> public class StrategyHelperTest extends NGCommonUtilitiesTestBase {
>   private static final String JSON_FOR_STEP_EXPANSION = "---\n"
> ```

can these files be file input?

File: 320-ci-execution/src/main/java/io/harness/ci/integrationstage/InitializeStepGenerator.java [hunk 1]
```diff
@@ -15,23 +15,33 @@
...
 
+  @Inject private StrategyHelper strategyHelper;
   InitializeStepInfo createInitializeStepInfo(ExecutionElementConfig executionElement, CodeBase ciCodebase,
       StageElementConfig stageElementConfig, CIExecutionArgs ciExecutionArgs, Infrastructure infrastructure,
       String accountId) {
     IntegrationStageConfig integrationStageConfig = IntegrationStageUtils.getIntegrationStageConfig(stageElementConfig);
 
+    List<ExecutionWrapperConfig> openedExecutionElement = new ArrayList<>();
+    for (ExecutionWrapperConfig config : executionElement.getSteps()) {
+      openedExecutionElement.addAll(strategyHelper.expandExecutionWrapperConfig(config));
```
Comment (harshjain12) [inline 320-ci-execution/src/main/java/io/harness/ci/integrationstage/InitializeStepGenerator.java:42]: add behind ff

**Review by harshjain12 — DISMISSED** [review]
> Reviewed 6 of 15 files at r1, 9 of 13 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34823)* status: 5 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by harshjain12 — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34823)* status: 4 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34823)* status: 4 file(s) to review.

**Pull request merged**